deepinterpolation/generator_collection.py

```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
 
-        self.total_frame_per_movie = int(raw_data.shape[0])
+        self.total_frame_per_movie = int(self.movie_data.shape[0])
 
         self._update_end_frame(self.total_frame_per_movie)
         self._calculate_list_samples(self.total_frame_per_movie)
 
-        average_nb_samples = np.min([int(raw_data.shape[0]), 100])
-
-        # For backward compatibility
-        if "cache_data" in self.json_data.keys():
-            self.cache_data = self.json_data["cache_data"]
-        else:
-            self.cache_data = False
-
-        if self.cache_data:
-            print('Caching hdf5 file... \n')
-            if self.end_frame > 0:
-                self.raw_data = raw_data[0:np.min([self.total_frame_per_movie,
-                    self.end_frame+self.post_frame+self.pre_post_omission+1]), :, :]
-            else:
-                self.raw_data = raw_data[:, :, :]
-            local_data = self.raw_data[0:average_nb_samples, :, :].flatten()
-        else:
-            local_data = raw_data[0:average_nb_samples, :, :].flatten()
-            self.cache_data = False
-
+        average_nb_samples = np.min(
+            [int(self.movie_data.shape[0]), self.movie_statistics_sample_size])
+        local_data = self.movie_data[:average_nb_samples].flatten()
         local_data = local_data.astype("float32")
 
-        self.local_mean = np.mean(local_data)
-        self.local_std = np.std(local_data)
+        self.local_mean = local_data.mean()
+        self.local_std = local_data.std()
+
+        if self._gpu_available:
+            self.cached_index = None
+            self._data_tensor = None
+            if not self.gpu_cache_full:
+                # Get indices for partial movie cache
+                start = self.pre_frame + self.pre_post_omission
+                end = start + self.batch_size
+                self.batch_indices = list(range(start,end))
+                self.input_indices = np.vstack(
+                    [self.__index_generation(frame_index) \
+                        for frame_index in self.batch_indices])
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
+        if self._movie_data is None:
+            with h5py.File(self.raw_data_file, "r") as movie_obj:
+                end_ind = self.end_frame+self.post_frame+self.pre_post_omission+1
+                total_frame_per_movie = movie_obj['data'].shape[0]
+                if self.end_frame > 0 and  total_frame_per_movie > end_ind:
+                    self._movie_data = movie_obj['data'][:end_ind]
+                else:
+                    self._movie_data = movie_obj['data'][()]
+            if self.normalize_cache:
+                self._movie_data = self._movie_data.astype("float32")
+                self._movie_data = _normalize(self._movie_data,
+                        self.local_mean,
+                        self.local_std)
+        if self.gpu_cache_full:
+            if self._data_tensor is None:
```

Comment:
I don't think you need this extra variable `self._data_tensor`. Can't you just convert `movie_data` to tensor and use `self.movie_data` throughout?